### PR TITLE
Fix Overscroll Behavior in Android 12+

### LINKED ISSUE #488: Android 12+: Stretch overscroll effect breaks schedule screen interaction
# Environment
- Android version: 12 (API 31) or newer
- Branch: `master`

# How to reproduce
1. Visit the schedule screen at the first column
2. Try to scroll before the first column (the column items stretch a little and the spacing on the right is lost)
3. Try to tap a session in the first column to open it

# Observed behavior
- The session won't open anymore.

  https://user-images.githubusercontent.com/144518/178827021-d241026c-cdb8-4488-8b16-1aa9a5dcb584.mp4

# Expected behavior
- The session details are shown.

# Workaround
- Rotate the device to re-render the schedule screen.

# Related
- [Android 12: Behavior changes / Stretch overscroll effect](https://developer.android.com/about/versions/12/behavior-changes-all#overscroll)

### PR DESCRIPTION
**Pull request by isiko**
# Description
As stated in #488 the overstretch animation of Android 12+ breaks when applied to the main Horizontal Scroll View. This should be fixed with this Pull request. The error was caused by the OnTouchListener catching OnTouch Events even though the target columnIndex was outside the allowed range.

# Before
When trying to scroll outside of the allowed horizontal range of the stretch Animation would not reset after releasing, making horizontal scrolling impossable.

# After
Horizontal scrolling functioning as with the old animation

Resolves #488

### CODE REVIEW

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollState.kt [hunk 1]
```diff
@@ -53,9 +53,11 @@ data class HorizontalSnapScrollState @JvmOverloads constructor(
 
     /**
      * Returns the column index calculated based on the given [scrollX] value
-     * and the current state of [HorizontalSnapScrollState].
+     * and the current state of [HorizontalSnapScrollState]. Normaly, the output
+     * Value is clamped to the allowed columnIndex range, which can be turned of
+     * using the [clamped] parameter
      */
-    fun calculateOnTouchColumnIndex(scrollX: Int): Int {
+    fun calculateOnTouchColumnIndex(scrollX: Int, clamped:Boolean = true): Int {
```
Comment (johnjohndoe) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollState.kt:60]: Why would we add the boolean flag as a public property if it is `false` in every use-case? This should be the only behavior. It looks like as if we can replace the min/max code here with the "limiting" code which is currently in `HorizontalSnapScrollView.OnTouchListener#onTouch`.

**Review by johnjohndoe — CHANGES_REQUESTED** [review]
Great work! This fix works fine - I tested it on a 10" tablet running Android 12.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollState.kt (hunk 1 above)
Comment (isiko) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollState.kt:60]: The the min/max Code was added with the Improvements made after transpiling the class from Java to Kotlin. I thought the lack of any range limitation was just a bug, as the function name does not imply that the output might not be a valid Index.
I thought it would be a good Idea to make the default behavior of the Method as safe/intuitive as possible, to avoid bugs produced in the future.

Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollState.kt:60]: I would not leave code/behavior behind (the suggested default behavior, `clamped: Boolean = true`) which is **not used** anywhere (at runtime). The purpose of the function is to calculate an index which is safe to be scrolled to. The range limitation can be calculated by the current state of properties in `HorizontalSnapScrollState` as you show in your pull request. Therefore, it is an **implementation detail** to the overall calculation and should be **hidden within** `calculateOnTouchColumnIndex()`. I would not split the index calculation into two methods - the caller would _need to know_ to invoke both which is considered bad interface design.

Comment (isiko) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollState.kt]: Updated the code, now the `calculateOnTouchColumnIndex()` returns -1 if the result would be invalid and the `OnTouch()` Method just checks if the returned Value is `>= 0`

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt [hunk 2]
```diff
@@ -173,9 +173,13 @@ class HorizontalSnapScrollView(context: Context?, attrs: AttributeSet?) : Horizo
                 return false
             } else if (event.action == ACTION_UP || event.action == ACTION_CANCEL) {
                 val scrollX = event.x.toInt()
-                val columnIndex = horizontalSnapScrollState.calculateOnTouchColumnIndex(scrollX)
-                scrollToColumn(columnIndex, fast = false)
-                return true
+
+                val columnIndexCalc = horizontalSnapScrollState.calculateOnTouchColumnIndex(scrollX)
+
+                if (columnIndexCalc >= 0) {
```
Comment (johnjohndoe) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:179]: How about `columnIndexCalc > -1`?

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt [hunk 3]
```diff
@@ -173,9 +173,13 @@ class HorizontalSnapScrollView(context: Context?, attrs: AttributeSet?) : Horizo
                 return false
             } else if (event.action == ACTION_UP || event.action == ACTION_CANCEL) {
                 val scrollX = event.x.toInt()
-                val columnIndex = horizontalSnapScrollState.calculateOnTouchColumnIndex(scrollX)
-                scrollToColumn(columnIndex, fast = false)
-                return true
+
+                val columnIndexCalc = horizontalSnapScrollState.calculateOnTouchColumnIndex(scrollX)
```
Comment (johnjohndoe) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:177]: How about `calculatedClumnIndex`?

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollStateTest.kt [hunk 4]
```diff
@@ -184,60 +184,60 @@ class HorizontalSnapScrollStateTest {
...
             createState().copy(
                 columnWidth = 993,
                 displayColumnCount = 4,
-                activeColumnIndex = 5,
+                activeColumnIndex = 1,
                 roomsCount = 10
             )
         ) {
-            assertThat(calculateOnTouchColumnIndex(scrollX = -columnWidth * 2)).isEqualTo(6)
+            assertThat(calculateOnTouchColumnIndex(scrollX = columnWidth * 2)).isEqualTo(-1)
```
Comment (johnjohndoe) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollStateTest.kt:240]: Why did you need to change the input values for these tests so dramatically? You could explain this in the commit message.

**Review by johnjohndoe — CHANGES_REQUESTED** [review]

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollState.kt (hunk 1 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollState.kt]: OK. I like this better.

**Comment by isiko** [comment]
Sorry for the double push, should be all fixed now :+1:

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollStateTest.kt (hunk 4 above)
Comment (isiko) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollStateTest.kt]: That wasn't intentional, although `calculateOnTouchColumnIndex returns -1 if the target index is higher than the max allowed in portrait mode to prevent scrolling off the screen to the right` seems have been faulty as it would not produce the edge case that it should

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollStateTest.kt [hunk 5]
```diff
@@ -208,27 +208,27 @@ class HorizontalSnapScrollStateTest {
...
                 columnWidth = 993,
                 displayColumnCount = 1,
-                activeColumnIndex = 4,
+                activeColumnIndex = 10,
                 roomsCount = 10
             )
         ) {
             val scrollThresholdWidth = (columnWidth * SCROLL_THRESHOLD_FACTOR).toInt()
-            assertThat(calculateOnTouchColumnIndex(scrollX = -scrollThresholdWidth + 1)).isEqualTo(4)
+            assertThat(calculateOnTouchColumnIndex(scrollX = -scrollThresholdWidth + 1)).isEqualTo(-1)
```
Comment (johnjohndoe) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollStateTest.kt:226]: Following up on my former comment. I wonder what `activeColumnIndex` is `10`. If I am not mistaken the column index is zero-based which means that the last index in this example would be `9`. With this value the test does not succeed. This again pointed me at the scroll calculation. Shouldn't it be `-scrollThresholdWidth - 1` instead of `-scrollThresholdWidth + 1`?

**Review by johnjohndoe — CHANGES_REQUESTED** [review]

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollStateTest.kt (hunk 5 above)
Comment (isiko) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollStateTest.kt]: Last Index is indeed 9, the test failed exactly for the reason you mentioned, it should be `-scrollThresholdWidth - 1` :facepalm:

**Review by johnjohndoe — APPROVED** [review]
Thank you for the fix!
